aws/resource_aws_kms_grant.go

```diff
@@ -0,0 +1,418 @@
+		Operations:       expandStringList(d.Get("operations").([]interface{})),
+	}
+
+	if v, ok := d.GetOk("name"); ok {
+		input.Name = aws.String(v.(string))
+	}
+	if v, ok := d.GetOk("constraints"); ok {
+		input.Constraints = expandKmsGrantConstraints(v.([]interface{}))
+	}
+	if v, ok := d.GetOk("retiring_principal"); ok {
+		input.RetiringPrincipal = aws.String(v.(string))
+	}
+	if v, ok := d.GetOk("grant_creation_tokens"); ok {
+		input.GrantTokens = expandStringList(v.([]interface{}))
+	}
+
+	log.Printf("[DEBUG]: Adding new KMS Grant: %s", input)
+
+	var out *kms.CreateGrantOutput
+
+	err := resource.Retry(2*time.Minute, func() *resource.RetryError {
+		var err error
+
+		out, err = conn.CreateGrant(&input)
+
+		if err != nil {
+			if awsErr, ok := err.(awserr.Error); ok {
+				// Error Codes: https://docs.aws.amazon.com/sdk-for-go/api/service/kms/#KMS.CreateGrant
+				// Under some circumstances a newly created IAM Role doesn't show up and causes
+				// an InvalidArnException to be thrown. TODO: Possibly change the aws_iam_role code?
+				if awsErr.Code() == "DependencyTimeoutException" ||
+					awsErr.Code() == "InternalException" ||
+					awsErr.Code() == "InvalidArnException" {
+					return resource.RetryableError(
+						fmt.Errorf("[WARN] Error adding new KMS Grant for key: %s, retrying %s",
+							*input.KeyId, err))
+				}
+			}
+			log.Printf("[ERROR] An error occured creating new AWS KMS Grant: %s", err)
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Created new KMS Grant: %s", *out.GrantId)
+	d.SetId(*out.GrantId)
+	d.Set("grant_id", out.GrantId)
+	d.Set("grant_token", out.GrantToken)
+
+	return resourceAwsKmsGrantRead(d, meta)
+}
+
+func resourceAwsKmsGrantRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).kmsconn
+
+	grantId := d.Id()
+	keyId := d.Get("key_id").(string)
+
+	log.Printf("[DEBUG] Looking for grant id: %s", grantId)
+	grant, err := findKmsGrantByIdWithRetry(conn, keyId, grantId)
+
+	if err != nil {
+		return err
+	}
+	if grant != nil {
+		// The grant sometimes contains principals that identified by their unique id: "AROAJYCVIVUZIMTXXXXX"
+		// instead of "arn:aws:...", in this case don't update the state file
+		if strings.HasPrefix(*grant.GranteePrincipal, "arn:aws") {
+			d.Set("grantee_principal", grant.GranteePrincipal)
+		} else {
+			log.Printf(
+				"[WARN] Unable to update grantee principal state %s for grant id %s for key id %s.",
+				*grant.GranteePrincipal, grantId, keyId)
+		}
+
+		if grant.RetiringPrincipal != nil {
+			if strings.HasPrefix(*grant.RetiringPrincipal, "arn:aws") {
+				d.Set("retiring_principal", grant.RetiringPrincipal)
+			} else {
+				log.Printf(
+					"[WARN] Unable to update retiring principal state %s for grant id %s for key id %s",
+					*grant.RetiringPrincipal, grantId, keyId)
+			}
+		}
+
+		d.Set("operations", grant.Operations)
+		if *grant.Name != "" {
```

Comment:
No, the grant name is optional, however if you don't set it AWS will set it to "". However, if I was to  load that into the state it would cause the state to mismatch the specified state which would cause resource re-creation. So if the grant name is empty just ignore it.